File: Pseudo-Palindromic_Paths_in_a_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int pseudoPalindromicPaths (TreeNode root) {
        return dfs(root, new HashSet<>());        
    }

    private int dfs(TreeNode node, Set<Integer> pairs) {
        if (node == null) {
            return 0;
        }

        if (pairs.contains(node.val)) {
            pairs.remove(node.val);
        } else {
            pairs.add(node.val);
        }

        if (node.left == null && node.right == null) {
            return pairs.size() <= 1 ? 1 : 0;
        }

        int left = dfs(node.left, new HashSet<>(pairs));
        int right = dfs(node.right, new HashSet<>(pairs));

        return left + right;
    }    
}

File: Pseudo-Palindromic_Paths_in_a_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:

        def dfs(node, pairs):
            if not node:
                return 0
            
            if node.val in pairs:
                pairs.remove(node.val)
            else:
                pairs.add(node.val)
            
            if not node.left and not node.right:
                return 1 if len(pairs) <= 1 else 0
            
            left = dfs(node.left, set(pairs))
            right = dfs(node.right, set(pairs))

            return left + right

        return dfs(root, set())
